fish-rust/src/fd_monitor.rs

```diff
@@ -0,0 +1,573 @@
+    /// is removed from the [`FdMonitor`] set.
+    callback: FdMonitorCallback,
+    /// The timeout associated with waiting on this item or `None` to wait indefinitely. A timeout
+    /// of `0` is not supported.
+    timeout: Option<Duration>,
+    /// The last time we were called or the time of initialization.
+    last_time: Option<Instant>,
+    /// The id for this item, assigned by [`FdMonitor`].
+    item_id: FdMonitorItemId,
+}
+
+/// Unlike C++, rust's `Vec` has `Vec::retain()` instead of `std::remove_if(...)` with the inverse
+/// logic. It's hard to keep track of which bool means what across the different layers, so be more
+/// explicit.
+#[derive(PartialEq, Eq)]
+enum ItemAction {
+    Remove,
+    Retain,
+}
+
+impl FdMonitorItem {
+    /// Return the duration until the timeout should trigger or `None`. A return of `0` means we are
+    /// at or past the timeout.
+    fn remaining_time(&self, now: &Instant) -> Option<Duration> {
+        let last_time = self.last_time.expect("Should always have a last_time!");
+        let timeout = self.timeout?;
+        assert!(now >= &last_time, "Steady clock went backwards or bug!");
+        let since = *now - last_time;
+        Some(if since >= timeout {
+            Duration::ZERO
+        } else {
+            timeout - since
+        })
+    }
+
+    /// Invoke this item's callback if its value (when its value is set in the fd or has timed out).
+    /// Returns `true` if the item should be retained or `false` if it should be removed from the
+    /// set.
+    fn service_item(&mut self, fds: &FdReadableSet, now: &Instant) -> ItemAction {
+        let mut result = ItemAction::Retain;
+        let readable = fds.test(self.fd.as_raw_fd());
+        let timed_out = !readable && self.remaining_time(now) == Some(Duration::ZERO);
+        if readable || timed_out {
+            self.last_time = Some(*now);
+            let reason = if readable {
+                ItemWakeReason::Readable
+            } else {
+                ItemWakeReason::Timeout
+            };
+            match &self.callback {
+                FdMonitorCallback::None => panic!("Callback not assigned!"),
+                FdMonitorCallback::Native(callback) => (callback)(&mut self.fd, reason),
+                FdMonitorCallback::Ffi(callback, param) => {
+                    // Safety: identical objects are generated on both sides by cxx bridge as
+                    // integers of the same size (minimum size to fit the enum).
+                    let reason = unsafe { std::mem::transmute(reason) };
+                    (callback)(&mut self.fd as *mut _, reason, *param)
+                }
+            }
+            if !self.fd.is_valid() {
+                result = ItemAction::Remove;
+            }
+        }
+        return result;
+    }
+
+    /// Invoke this item's callback with a poke, if its id is present in the sorted poke list.
+    fn maybe_poke_item(&mut self, pokelist: &[FdMonitorItemId]) -> ItemAction {
```

Comment:
ideally the rename from `poke_item()` to `maybe_poke_item()` should be in a separate commit unless it's related to the port?

Reply:
Fair enough.